src/demand.c

```diff
@@ -66,68 +42,102 @@ demand_data_t *create_demand_data(double base_demand, int pattern_index, const c
 }
 
 void delete_demand_data(void *data)
-{ 
+{
     demand_data_t *demand_data = *(demand_data_t **)data;
 
     if (demand_data->category_name)
         free(demand_data->category_name);
 
-	free(demand_data);
+    free(demand_data);
 }
 
-size_t get_demand_data_size(void)
+// ---  Exported functions  --- //
+
+int create_demand_list(list_t **dlist)
 {
-	return sizeof(demand_data_t *);
+    list_t *demand_list = create_list(sizeof(demand_data_t *), delete_demand_data);
+	if (!demand_list) return 0;
+	*dlist = demand_list;
+    return 1;
 }
 
-demand_data_t *get_demand_data(list_node_t *lnode)
+void delete_demand_list(list_t **dlist)
 {
-    return *(demand_data_t **)get_data(lnode);
+    if (*dlist) delete_list(*dlist);
+    *dlist = NULL;
 }
 
-
-bool convert_units(list_node_t *lnode, double unit_conversion)
+int add_demand(list_t *dlist, double baseDemand, int pattern, const char *name)
 {
-	double base_demand = get_base_demand(lnode);
-
-	set_base_demand(lnode, base_demand/unit_conversion);
-	return true;
+    demand_data_t *demand_data;
+    demand_data = create_demand_data(baseDemand, pattern, name);
+    return append_list(dlist, &demand_data);
 }
 
-
-double get_base_demand(list_node_t *lnode)
+void remove_nth_demand(list_t *dlist, int n)
 {
-	return get_demand_data(lnode)->base_demand;
+    remove_nth_node(dlist, n);
 }
 
-void set_base_demand(list_node_t *lnode, double base_demand)
+int get_num_demands(list_t *dlist)
 {
-	get_demand_data(lnode)->base_demand = base_demand;
+    return size_list(dlist);
 }
 
-int get_pattern_index(list_node_t *lnode)
+int get_demand_index(list_t *dlist, const char *name, int key)
 {
-    return get_demand_data(lnode)->pattern_index;
+    demand_data_t *dd;
+    int index = 1;
+
+    // Search for key
+    if (name == NULL || strlen(name) == 0)
+    {
+        return get_index(dlist, key);
+    }
+
+    // Search for category name
+    dd = get_first_demand(dlist);
+    while (dd)
+    {
+        if (strcmp(dd->category_name, name) == 0) return index;
+        index++;
+        dd = get_next_demand(dlist);
+    }
+    return 0;
 }
 
-void set_pattern_index(list_node_t *lnode, int pattern_index)
+demand_data_t *get_first_demand(list_t *dlist)
```

Comment:
The way that demands were handled originally was in an explicit list-like manner and you objected to it. I'm trying to avoid list clients having to know about "nodes" for no good reason. It's much cleaner to provide direct access to the data that the list stores, rather than its nodes (as was done in the original generic list code).